prusti-interface/src/environment/mod.rs

```diff
@@ -200,4 +201,21 @@ impl<'r, 'a, 'tcx> Environment<'r, 'a, 'tcx> {
     pub fn get_procedure(&self, proc_def_id: ProcedureDefId) -> Procedure<'a, 'tcx> {
         Procedure::new(self.tcx(), proc_def_id)
     }
+
+    /// Get an associated item by name.
+    pub fn get_assoc_item(&self, id: DefId, name: Symbol) -> Option<ty::AssociatedItem> {
+        self.tcx().associated_items(id).find(|assoc_item| assoc_item.name == name)
+    }
+
+    /// Get a trait method declaration by name for type.
+    pub fn get_trait_method_decl_for_type(&self, typ: ty::Ty<'tcx>, trait_id: DefId, name: Symbol) -> Option<ty::AssociatedItem> {
+        let mut result = None;
+        self.tcx().for_each_relevant_impl(trait_id, typ, |impl_id| {
+            let item = self.get_assoc_item(impl_id, name);
+            if item.is_some() {
+                result = item;
```

Comment:
Last thing: could you `assert!(result.is_none())` in this branch?

Reply:
Sure will, do. I will also try to change this PR to not have the above mentioned restriction. I will have an in depth look into this next week.

Reply:
Which restriction? The `assert!(result.is_none())` would not be a negative thing in my opinion.

Reply:
The restriction with the `self` parameter. The current solution does not work for functions without an implicit `self` type.

Reply:
I did not add `assert!(result.is_none())` as the code currently has issues differentiating between generic traits implemented with different associated types. See my new comment below for an example. Putting this assert here would create a panic in such situations, even though they can be handled just as they were before; i.e. we simply do not support trait refinement checking on such traits if several implementations exist for a single type.

Reply:
I see that you already changed the return type to `Result<..>`; perfect! I was about to ask that.